### PR TITLE
[40031] Left-side search pane to add existing work packages to team planner

### PR DESCRIPTION
**Pull request by HDinger**
### Requirements

**Tl;dr**
* Show a pane on the left of the team planner to add existing wps to the team planner
* The pane is searchable
* The results are displayed as cards and can be moved to the team planner via drag&drop

**Displaying the left-side search pane**
- [x] There is a button on calendars toolbar on the left, with the title "+ Add existing".
- [x] The button is a toggle
  - [x] clicking on it activates the add existing search pane on the left-hand side of the screen, and the button state changes to active.
  - [x] Clicking on the button when active then hides the left-side pane, and returns the button to its default (non-active) state.
- [x] The left pane takes the entire available vertical height 
- [x] The pane can overflow; users can scroll within it when there are a lot of search results.

**Empty state**

- [x] The left pane is initially empty
- [x] The empty state consists of a search box on the top, an illustration and a text explanation 

**Searching**

- [x] Entering a search term begins the search process. The search mechanism is the same as that of the existing main search bar.
- [x] The scope is nevertheless limited by any active filters on the team planner.
- [x] Whilst the results are loading, an animated skeletal outline of a card serves as a loading indicator

**Search results**

- [x] Once the results are available, they are displayed as cards, one per work package, sorted by ID in descending order
- [x] Each card contains display attributes
  - [x] Line 1: Project name (truncated after 10 characters), Type, Status (truncated if necessary),
  - [x] Line 2: Title
  - [x] Line 3 (optional, only if information exists): Assignee (left-side, truncated after 15 characters), Start/end-dates (right side, format "Aug 2–5, 2021" for brevity)
  - [x] Overdue work packages (end date is in the past) display the dates in red.
  - [x] On boards: Hover displays the icons (close and icon) on the left instead, covering the project bit entirely. The "info" icon is blue and "remove" is red.
- [x] Work Packages currently shown in the team planner, won't be part of the search results

**Adding to team planner**

- [x] To add a work package to the team planner, the user may simply drag and drop it on to a calendar date.
- [x] During drag, calendar dates turn blue to indicate currently selected date on hover.
- [x] The space where the work package was dragged is represented by a dashed outline (the same as in the boards).
- [x] The number of calendar days highlighted on drag depends on the duration of the work package, inferred from start/end dates if present:
  - [x] If start/end dates are known, drag highlights the number of calendar days corresponding to the duration of the work package, in blue. (Within the limit of the current view, naturally)
  - [x] If start/end dates are unknown, only one day is highligted.
- [x] On mouse out, the work package is dropped onto the calendar. The work package is then assigned to the assignee of the row it was dropped on, and for the start date that was selected.
  - [x] If the work package previously did not have start/end dates, initially the start date will be the same as the end date (a duration of 1 day).
  - [x] If the work package already had an end date, that duration (inferred) is maintained, and an appropriate end date automatically selected.

**Technical**

- [x] Make newly dropped events clickable to open the split screen
- [x] After dropping the wp, it is removed from the short list
- [ ] Handle special error cases (e.g. start and end date cannot be changed due to inheritance)
- [ ] **Blocked** The button is only visible if the user has the right to manage team planners (depends on #10061 )
- [x] Avoid overlapping of avatar name and date in cards (awaits clarification by product team)
- [x] Write tests for new functionality

### Out of scope
* Removing a work package from the team planner
* The user can cancel the drag at any point by either pressing the escape key or dragging the work package back to the search pane.

[OP#40031](https://community.openproject.org/projects/openproject/work_packages/40031/activity)

### CODE REVIEW

**Comment by oliverguenther** [comment]
- [ ] The filter should receive a debounce, as right now it's triggering a search for every key added

- [ ] Dragging a card onto a resource with an existing element puts it on top (this may be unavoidable fullcalendar behavior) 

Peek 2022-01-24 08-27

> If the work package already had an end date, that duration (inferred) is maintained, and an appropriate end date automatically selected.
- ~~This does not appear to work, neither with manual nor automatic scheduling mode~~

File: frontend/src/app/features/team-planner/team-planner/add-work-packages/add-existing-pane.component.ts [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+export class AddExistingPaneComponent extends UntilDestroyedMixin {
+  @ViewChild('container') container:ElementRef;
+
+  @ViewChild('container')
+  set dragContainer(v:ElementRef|undefined) {
+    // ViewChild reference may be undefined initially
+    // due to ngIf
+    if (v !== undefined) {
+      this.calendarDrag.destroyDrake();
+      this.calendarDrag.registerDrag(v, '.op-add-existing-pane--wp');
```
Comment (oliverguenther) [inline frontend/src/app/features/team-planner/team-planner/add-work-packages/add-existing-pane.component.ts:53]: The CalendarDragDropService is provided in root, so this will break if there is anything going on between two instances accessing the drag service. This is unlikely but for other implementations, we have chosen to provide services with stateful actions (like an instance to the drake) locally. I've seen you also use it in team-planner.component, so providing it in the page might make sense.

Comment (oliverguenther) [inline frontend/src/app/features/team-planner/team-planner/add-work-packages/add-existing-pane.component.ts:53]: The CalendarDragDropService is provided in root, so this will break if there is anything going on between two instances accessing the drag service. This is unlikely but for other implementations, we have chosen to provide services with stateful actions (like an instance to the drake) locally. I've seen you also use it in team-planner.component, so providing it in the page might make sense.

File: frontend/src/app/features/team-planner/team-planner/add-work-packages/add-existing-pane.component.ts [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+      const currentFilters = this.urlParamsHelper.buildV3GetFilters(query.filters);
+
+      currentFilters.forEach((filter) => {
+        Object.keys(filter).forEach((name) => {
+          if (name !== 'assignee' && name !== 'datesInterval') {
+            filters.add(name, filter[name].operator, filter[name].values);
+          }
+        });
+      });
+    }
```
Comment (oliverguenther) [inline frontend/src/app/features/team-planner/team-planner/add-work-packages/add-existing-pane.component.ts]: I'd probably extract those into a separate method, as `searchWorkPackages` is becoming quite unwieldy

File: frontend/src/app/features/team-planner/team-planner/add-work-packages/add-existing-pane.component.ts [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+
+    this
+      .apiV3Service
+      .withOptionalProject(this.currentProject.id)
+      .work_packages
+      .filtered(filters)
+      .get()
+      .pipe(
+        debounceTime(100),
```
Comment (oliverguenther) [inline frontend/src/app/features/team-planner/team-planner/add-work-packages/add-existing-pane.component.ts]: I see you've added a debounce, but this does not work as you're creating a new observable everytime. You need to pipe the input to a switchMap with a debounce in between to make it work. Let me know if you want any help there.

**Review by oliverguenther — CHANGES_REQUESTED** [review]
Mostly looking good to me :+1:  Some of the spec items are not working as described, and some minor code feedback.

File: frontend/src/app/features/team-planner/team-planner/add-work-packages/add-existing-pane.component.html [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+  <op-icon
+    class="op-add-existing-pane--search-icon"
+    icon-classes="icon-context icon-search">
+  </op-icon>
+
+  <button
+    *ngIf="isSearching"
+    (click)="clearInput()"
+    class="op-add-existing-pane--close-search"
```
Comment (b12f) [inline frontend/src/app/features/team-planner/team-planner/add-work-packages/add-existing-pane.component.html:19]: This should be a `type=button` so we don't accidentally submit an outer form.

File: frontend/src/app/features/team-planner/team-planner/add-work-packages/add-existing-pane.component.ts [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+import { UntilDestroyedMixin } from 'core-app/shared/helpers/angular/until-destroyed.mixin';
+import { CalendarDragDropService } from 'core-app/features/team-planner/team-planner/calendar-drag-drop.service';
+
+@Component({
+  selector: 'op-add-existing-pane',
+  templateUrl: './add-existing-pane.component.html',
+  styleUrls: ['./add-existing-pane.component.sass'],
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class AddExistingPaneComponent extends UntilDestroyedMixin {
```
Comment (b12f) [inline frontend/src/app/features/team-planner/team-planner/add-work-packages/add-existing-pane.component.ts]: Perhaps add the `op-add-existing-pane` base BEM class via `@HostBinding` just to be correct in BEM application. Unfortunately angulars styling implementation doesn't allow us to select it, but I think it makes the rendered HTML more readable if it's clear what the base block is.

File: frontend/src/app/features/team-planner/team-planner/calendar-drag-drop.service.ts [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+
+    if (!isElementStillVisible) {
+      this.draggableWorkPackages$
+        .next(oldDraggables.concat(workPackage));
+    }
+  }
+
+  private eventData(eventEl:HTMLElement):undefined|DragMetaInput {
+    const wpID = eventEl.dataset.dragHelperId;
+    if (wpID) {
```
Comment (b12f) [inline frontend/src/app/features/team-planner/team-planner/calendar-drag-drop.service.ts]: I'd invert these checks;

```
if (!wpID) {
  return;
}
```

Reduces nesting, making hte code more readable imo

File: frontend/src/app/features/work-packages/components/wp-buttons/wp-status-button/wp-status-button.component.sass [hunk 7]
```diff
@@ -13,3 +13,11 @@
 
   &--text
     @include text-shortener
+
+  &_small
+    .op-wp-status-button--button
```
Comment (b12f) [inline frontend/src/app/features/work-packages/components/wp-buttons/wp-status-button/wp-status-button.component.sass]: Could be shortened to `&_small &--button`

**Review by b12f — COMMENTED** [review]
Some nitpicks

**Comment by HDinger** [comment]
> * Dragging a card onto a resource with an existing element puts it on top (this may be unavoidable fullcalendar behavior)

We could disable that via a fullCalendar Option. However, when there is no white space in an assignee row (see picture below), the element cannot be dropped (because overlapping is disabled). I'd therefore leave it like it is for now, until we find a smart solution.
Bildschirmfoto 2022-01-24 um 12 01 49

**Review by oliverguenther — APPROVED** [review]
